Implement patch control points sample

Zbigniew-Jedlinski-Mobica opened the pull request:
## Description
Demonstrate how to use patch control points in extended dynamic state 2, which allows dynamic change of patch control points in tessellation.
Sample requires new model in assets: https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/21

## Notes
Sample was tested using following configurations:

Windows 11 Education 22H2
NVIDIA GeForce RTX 3080 Ti (NVIDIA driver version: 31.0.15.3118, driver date: 24.02.2023)
Intel(R) Core(TM) i7-9700K
16GB RAM

Windows 10 Pro
Nvidia T600 Laptop GPU (NVIDIA driver version: 31.0.15.2849, driver date: 02.02.2023)
Intel i7-11850H
32GB RAM

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

Zbigniew-Jedlinski-Mobica committed:
Implement patch control points sample (#2)


Co-authored-by: Karina Murawko-Wiśniewska <karina.murawko-wisniewska@mobica.com>

SaschaWillems commented:
The sample works fine for me on Win10 with an NV RT 2060.

A few things I noticed during my review:

- Camera rotation on the horizontal axis is inverted (compared to all other samples)
- Tessellation should be enabled by default
- Default camera position is so far away, that tessellation is not visible until you move closer to the terrain. Default camera position should be moved closer.
- What is the difference between the two terrain patches? Both look the same, and both change their tessellation when moving the slider. Shouldn't the sample demonstrate a difference between the static and dynamic setup?

A minor note on naming: I'm not a native speaker, but afaik it should be  `dynamic_tessellation` instead of `dynamically_tessellation`. Some for `statically_tessellation`.

Zbigniew-Jedlinski-Mobica committed:
Code format fix in patch_control_points.h

According to Clang format checker updated code format

Patryk-Jastrzebski-Mobica committed:
Added colors depending on the level of tesselation and rotated objects (#3)

Zbigniew-Jedlinski-Mobica committed:
Merge branch 'KhronosGroup:main' into main

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=649) <br/>All committers have signed the CLA.

Zbigniew-Jedlinski-Mobica committed:
Updating the sample as commented in the review

Image of sample updated
Camera position updated
Tessellation enabled after running sample
Updated dynamically_tessellation and statically_tessellation names
Asset updated

Zbigniew-Jedlinski-Mobica committed:
Code format fix

According to Clang format checker updated code format

Zbigniew-Jedlinski-Mobica commented:
> 

Camera rotation is fixed.
Tesselation is enable by default.
Camera position now is closer to terrain model.
Terrain patches have pipelines built in different ways. Static is built without using patch control points. Dynamic is built using patch control points that can be modified in the cpp file.

Zbigniew-Jedlinski-Mobica commented:
@SaschaWillems , does it look fine now after the changes?

asuessenbach commented on samples/extensions/patch_control_points/README.md:
```diff
@@ -0,0 +1,148 @@
+  
+| Static/Non-dynamic | Dynamic State 2 |
+| ------------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------- |
+| dynamic_state = \{} | dynamic_state = {VK_DYNAMIC_STATE_VIEWPORT,<br>VK_DYNAMIC_STATE_SCISSOR,<br>**VK_DYNAMIC_STATE_PATCH_CONTROL_POINTS_EXT**} |
```
List either VIEWPORT and SCISSOR here on both sides or on neither.

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on samples/extensions/patch_control_points/README.md:
```diff
@@ -0,0 +1,148 @@
+graphics_create.pInputAssemblyState = &input_assembly_state;
+graphics_create.pDynamicState       = &dynamic_state;
+graphics_create.layout              = pipeline_layouts.dynamically_tessellation;
+...
```
Maybe note, that you still have to set `graphics_create.pTessellationState` to point to some valid `VkPipelineTessellationStateCreateInfo`

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on samples/extensions/patch_control_points/patch_control_points.cpp:
```diff
@@ -0,0 +1,611 @@
+	ubo_tess.tessellation_factor = gui_settings.tess_factor;
+	if (!gui_settings.tessellation)
+	{
+		// Setting this to zero sets all tessellation factors to 1.0 in the shader
```
factors -> levels

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on samples/extensions/patch_control_points/patch_control_points.cpp:
```diff
@@ -0,0 +1,611 @@
+	uniform_buffers.dynamic_tessellation->convert_and_update(ubo_tess);
+
+	/* Statically tessellation */
+	uniform_buffers.static_tessellation->convert_and_update(ubo_tess);
```
Do the dynamic_tessellation and the static_tessellation UBs ever differ?
If not, why do you have two of them?

Zbigniew-Jedlinski-Mobica replied:
static_tessellation is for drawing the model without using the patch control points functionality while dynamic_tessellation uses it. Two pipelines are used here to show that the same effect can be achieved with patch control points

asuessenbach commented on samples/extensions/patch_control_points/patch_control_points.cpp:
```diff
@@ -0,0 +1,611 @@
+	                                &pipeline_layouts.static_tessellation));
+
+	/* Second descriptor set */
+	set_layout_bindings = {
```
Is there any difference to what set_layout_bindings was before?

Zbigniew-Jedlinski-Mobica replied:
The set layout is set the same for static_tessellation and dynamic_tessellation

asuessenbach replied:
Why do you initialize it again, then?

Zbigniew-Jedlinski-Mobica replied:
> Why do you initialize it again, then?

There are two initializations to make it more visible that one is used for the static pipeline and the other for the dynamic pipeline.

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+layout(location = 1) in vec3 inNormal[];
+
+layout(location = 0) out vec3 outPos[3];
+layout(location = 1) out vec3 outNormal[3];
```
outPos and outNormal are not used in tess.tese, so no need to pass them around.
outColor could then be moved to location = 0.

Zbigniew-Jedlinski-Mobica replied:
Removed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+
+float GetTessLevel(vec4 p0, vec4 p1)
+{
+	float tesselationValue = 0.0f;
```
tessellationValue -> tessellationLevel ?

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+{
+	float tesselationValue = 0.0f;
+	float midDistance = 1.6f;
+	float shortDistance = 1.0f;
```
make midDistance and shortDistance const ?

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+	vec4 vCamView = ubo.projection * vCam;
+
+	// Calculate the vector length
+	float AvgDistance = length(vCamView);
```
Maybe AvgDistance -> centerDistance (and midPoint -> centerPoint) ?

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+	{
+		gl_TessLevelOuter[0] = GetTessLevel(gl_in[2].gl_Position, gl_in[0].gl_Position);
+		gl_TessLevelOuter[1] = GetTessLevel(gl_in[0].gl_Position, gl_in[1].gl_Position);
+		gl_TessLevelOuter[2] = GetTessLevel(gl_in[1].gl_Position, gl_in[2].gl_Position);
```
outColor is calculated three times here!
Maybe introduce some getColor() function, determining the color depending on the center of the triangle?

Zbigniew-Jedlinski-Mobica replied:
Function to get color was added to tesc file

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+layout(location = 1) out vec3 outNormal[3];
+layout(location = 2) out vec3 outColor[3];
+
+float GetTessLevel(vec4 p0, vec4 p1)
```
GetTessLevel -> getTessLevel ?

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -0,0 +1,100 @@
+layout(vertices = 3) out;
+
+layout(location = 0) in vec3 inPos[];
+layout(location = 1) in vec3 inNormal[];
```
inPos and inNormal are not used here, can be removed.

Zbigniew-Jedlinski-Mobica replied:
Removed

asuessenbach commented on shaders/patch_control_points/tess.vert:
```diff
@@ -0,0 +1,50 @@
+layout(location = 1) in vec3 inNormal;
+
+layout(location = 0) out vec3 outPos;
+layout(location = 1) out vec3 outNormal;
```
outPos and outNormal are not used in tess.tec, so no need to calculate and pass them around here.

Zbigniew-Jedlinski-Mobica replied:
Removed

asuessenbach commented on samples/extensions/patch_control_points/patch_control_points.cpp:
```diff
@@ -0,0 +1,611 @@
+	/* Attribute descriptions */
+	std::vector<VkVertexInputAttributeDescription> vertex_input_attributes = {
+	    vkb::initializers::vertex_input_attribute_description(0, 0, VK_FORMAT_R32G32B32_SFLOAT, 0),                               // Position
+	    vkb::initializers::vertex_input_attribute_description(0, 1, VK_FORMAT_R32G32B32_SFLOAT, offsetof(Vertex, normal)),        // Normal
```
Normal is not used in tess.vert, so need to describe it here.

Zbigniew-Jedlinski-Mobica replied:
Normal was removed from code

asuessenbach reviewed (commented):
In a sample on dynamic usage of tessellation patch control points I would expect some dynamic usage of them. That is, maybe, two or more models with varying numbers of patch control points.
Besides that, it looks really good!
Just a couple of mostly minor issues...

Zbigniew-Jedlinski-Mobica committed:
Updating the sample according to Andreas suggestions

Readme update
gui_settings.tess name update
ubo_tess.tessellation name update
shaders update

Zbigniew-Jedlinski-Mobica committed:
Merge branch 'KhronosGroup:main' into main

Zbigniew-Jedlinski-Mobica commented:
@asuessenbach when I loaded models with different number of control points in one pipeline, they were displayed incorrectly.
Each pipeline has a defined shader file that supports a certain number of patch control points.
As I understand this behavior is related to the way GLSL works

asuessenbach commented on samples/extensions/patch_control_points/patch_control_points.cpp:
```diff
@@ -145,11 +145,11 @@ void PatchControlPoints::update_uniform_buffers()
 	{
-		// Setting this to zero sets all tessellation factors to 1.0 in the shader
-		ubo_tess.tessellation_factor = 0.0f;
+		// Setting this to zero sets all tessellation levels to 1.0 in the shader
```
Is this comment still valid?

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -31,58 +31,69 @@ ubo_tessellation;
 	{
-		tesselationValue = ubo_tessellation.tessellationFactor * 0.4;
-		outColor[gl_InvocationID] = vec3(0.0f, 0.0f, 1.0f); // blue color
+		tessellationLevel = ubo_tessellation.tessellationFactor * 0.4;
```
0.4 -> 0.4f

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.tesc:
```diff
@@ -31,58 +31,69 @@ ubo_tessellation;
+	{
+		outColor = vec3(1.0f, 0.0f, 0.0f);        // red color
+	}
+	else if (tessellationLevel == ubo_tessellation.tessellationFactor * 0.4)
```
0.4 -> 0.4f

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented on shaders/patch_control_points/tess.vert:
```diff
@@ -35,16 +31,13 @@ layout(push_constant) uniform Push_Constants
+	rotX[1][2]  = sin(PI);
+	rotX[2][1]  = -sin(PI);
+	rotX[2][2]  = cos(PI);
+	vec4 pos    = rotX * vec4(inPos + push_constants.direction, 1.0f);
```
skip `pos` and directly assign the result to `gl_Position`?

Zbigniew-Jedlinski-Mobica replied:
Fixed

asuessenbach commented:
> Each pipeline has a defined shader file that supports a certain number of patch control points.

Then why use a dynamic number of patch control points?

Zbigniew-Jedlinski-Mobica commented:
> Then why use a dynamic number of patch control points?

The use of a dynamic number of patch control points is to cover the use of dynamic state 2. Dynamic state 1, 2 and 3 provide the functionality of using a dynamic pipeline instead of a static one.

Zbigniew-Jedlinski-Mobica committed:
Merge branch 'KhronosGroup:main' into main

Zbigniew-Jedlinski-Mobica committed:
Updating the sample according to Andreas suggestions

Updated typos
Corrected errors with types

Zbigniew-Jedlinski-Mobica committed:
Merge branch 'main' into main

Zbigniew-Jedlinski-Mobica commented:
> Why do you initialize it again, then?

There are two initializations to make it more visible that one is used for the static pipeline and the other for the dynamic pipeline.

JoseEmilio-ARM commented:
Tested on Android (Pixel 6), where extension is supported and sample seems to run fine, but the results do not look as expected e.g. on Windows. Maybe it is just wireframe not working as expected, but also increasing tessellation level does not have an effect. 

![Screenshot_20230703-150544](https://github.com/KhronosGroup/Vulkan-Samples/assets/33630720/23bdefb2-4e6a-41f9-a034-af23b9685c8e)

Code looks good however, approving. Android not a blocker as this extension might have limited use on mobile.
